### ISSUE TITLE
Potential memory leak using the tusd library

### ISSUE THREAD

**Issue by jdcalvin** [tus/tusd#876]
**Question**
I'm using the tusd library in a program structured similarly to the example here: https://github.com/tus/tusd/blob/master/docs/usage-package.md

It's running in production on a kubernetes cluster. The server currently receives about 75k upload POST requests daily on weekdays, and a varying amount of PATCH requests depending on upload size. 

Memory usage is overall really good considering the traffic, however we have what appears to be a memory leak that correlates to requests:
<img width="539" alt="image" src="https://user-images.githubusercontent.com/5394560/211894530-0b9e384c-d5a1-4fe5-abf3-e96d5affa0c6.png">
Requests:
<img width="518" alt="image" src="https://user-images.githubusercontent.com/5394560/211894603-672b8e6c-b441-4c32-913e-60b96af70d9a.png">
Instances restart on a fairly regular schedule so it's not a big problem, but left alone I'd expect to see the memory increase indefinitely.

We have the `CompleteUploads` channel receiving events very similar to the example above. Looking at the [tusd unrouted handler code](https://github.com/tus/tusd/blob/master/pkg/handler/unrouted_handler.go#L396) I don't see where the channel is explicitly closed, and wondering if it could be related to that.

**Setup details**
Please provide following details, if applicable to your situation:
- Operating System: alpine:3.16.2
- Used tusd version: 0.10.0
- Used tusd data storage: s3store
- Used tusd configuration: n/a
- Used tus client library: tus-js-client

**Comment by Acconut** [comment]
This does indeed look like a memory leak, but I am unsure if it is inside tusd. We have been running tusd as the binary (not the package) for many years with S3 and never experienced a memory leak. So I could imagine that there is another part of your application that is causing a growing memory usage. Maybe even the way how you consume hooks from tusd is problematic (and the suggestions in https://github.com/tus/tusd/blob/master/docs/usage-package.md are faulty). Could you try to profile your program and see what is causing the growth?

**Comment by jdcalvin** [comment]
We're only handling the Complete Uploads hook via `NotifyCompleteUploads` I did remove it and trying out an alternative approach. Will monitor to see how memory is impacted. I do plan on spending some more time with memory profiling. A few small attempts didn't show anything interesting though.

**Comment by Acconut** [comment]
Let us know if you find anything. Go's pprof has been really helpful whenever I used it.

> Looking at the [tusd unrouted handler code](https://github.com/tus/tusd/blob/master/pkg/handler/unrouted_handler.go#L396) I don't see where the channel is explicitly closed, and wondering if it could be related to that.

I doubt that. The linked channel exists once for each tusd instance and not for every upload. As such, the one channel never gets closed and should not contribute to memory growth.

**Comment by jdcalvin** [comment]
I've spent some time playing around with pprof but didn't get any positive signal for memory leaks. Same results from looking at `docker stats.` 

I refactored our code to not rely on hooks and confirmed that did not make any impact on memory usage. I haven't ruled out issues with kubernetes releasing memory so thats what I'll try next when I have time. I'll close this out for now since its very likely unrelated to the tusd package.
